cmd/main.go

```diff
@@ -30,9 +30,14 @@ import (
 // global flags for minio.
 var globalFlags = []cli.Flag{
 	cli.StringFlag{
-		Name:  "config-dir, C",
-		Value: getConfigDir(),
```

Comment:
This should not be removed

Reply:
If it is not removed, then the logic in `handleCommonCmdArgs` needs to be changed.

Reply:
https://github.com/minio/minio/pull/5486#issuecomment-368815261 is working for all the cases.  What are changes required to work for what case?

Reply:
In the following code, in normal case, the code execution will not reach the line `configDir = ctx.Parent().GlobalString("config-dir")`
```} else if ctx.GlobalIsSet("config-dir") {
		configDir = ctx.GlobalString("config-dir")
		// cli package does not expose parent's "config-dir" option.  Below code is workaround.
		if configDir == "" {
			configDir = ctx.Parent().GlobalString("config-dir")
		}
	} else {```

Reply:
This can be verified when you try `minio -C /tmp/configDir gateway s3`

Reply:
I tried below and everything works as expected.
```bash
$ MINIO_ACCESS_KEY=minio MINIO_SECRET_KEY=minio123 ./minio gateway s3
Created minio configuration file successfully at /home/bala/.minio

$  MINIO_ACCESS_KEY=minio MINIO_SECRET_KEY=minio123 ./minio gateway s3 -C /tmp/c1
Created minio configuration file successfully at /tmp/c1

$ MINIO_ACCESS_KEY=minio MINIO_SECRET_KEY=minio123 ./minio gateway -C /tmp/c2 s3
Created minio configuration file successfully at /tmp/c2

$ MINIO_ACCESS_KEY=minio MINIO_SECRET_KEY=minio123 ./minio -C /tmp/c3 gateway s3
Created minio configuration file successfully at /tmp/c3
```

Why are you saying its not working?  Have you tested it?

Reply:
I am pasting the test that I did for reference
With changes suggested, a folder called `/tmp/test` doesn't exist on my system. I created a buggy config.json inmy default folder. As can be seen from the output, the user specified folder was not created.
```
 kannappan@x1minio ~/gopath/src/github.com/minio/minio (config_dir) $ git diff
diff --git a/cmd/main.go b/cmd/main.go
index d90813b1..e2a5a9e0 100644
--- a/cmd/main.go
+++ b/cmd/main.go
@@ -30,7 +30,8 @@ import (
 // global flags for minio.
 var globalFlags = []cli.Flag{
        cli.StringFlag{
-               Name: "config-dir, C",
+               Name:  "config-dir, C",
+               Value: getConfigDir(),
                Usage: func() string {
                        usage := "Path to configuration directory."
                        if getConfigDir() == "" {
kannappan@x1minio ~/gopath/src/github.com/minio/minio (config_dir) $ make
Building minio binary: /home/kannappan/gopath/src/github.com/minio/minio/minio
kannappan@x1minio ~/gopath/src/github.com/minio/minio (config_dir) $ ./minio -C /tmp/test gateway s3

Trace: 1: cmd/common-main.go:46:cmd.initConfig()
       2: cmd/gateway-main.go:150:cmd.StartGateway()
       3: cmd/gateway/s3/gateway-s3.go:89:s3.s3GatewayMain()
       4: vendor/github.com/minio/cli/app.go:499:cli.HandleAction()
       5: vendor/github.com/minio/cli/command.go:214:cli.Command.Run()
       6: vendor/github.com/minio/cli/app.go:386:cli.(*App).RunAsSubcommand()
       7: vendor/github.com/minio/cli/command.go:302:cli.Command.startApp()
       8: vendor/github.com/minio/cli/command.go:100:cli.Command.Run()
       9: vendor/github.com/minio/cli/app.go:260:cli.(*App).Run()
      10: cmd/main.go:155:cmd.Main()
      11: main.go:71:main.main()
[2018-02-28T08:25:02.6446234Z] [FATAL] Unable to load config version: '22'. (configuration version mismatch. Expected: '22', Got: '32')
kannappan@x1minio ~/gopath/src/github.com/minio/minio (config_dir) $ 
```

Reply:
Without any changes, the folder was created.

```
 kannappan@x1minio ~/gopath/src/github.com/minio/minio (config_dir) $ git diff
kannappan@x1minio ~/gopath/src/github.com/minio/minio (config_dir) $ make
Building minio binary: /home/kannappan/gopath/src/github.com/minio/minio/minio
kannappan@x1minio ~/gopath/src/github.com/minio/minio (config_dir) $ ./minio -C /tmp/test gateway s3
Created minio configuration file successfully at /tmp/test


               *** Warning: Not Ready for Production ***

Endpoint:  http://192.168.86.26:9000  http://172.17.0.1:9000  http://127.0.0.1:9000
AccessKey: minio 
SecretKey: minio123 

Browser Access:
   http://192.168.86.26:9000  http://172.17.0.1:9000  http://127.0.0.1:9000

Command-line Access: https://docs.minio.io/docs/minio-client-quickstart-guide
   $ mc config host add mys3 http://192.168.86.26:9000 minio minio123

Object API (Amazon S3 compatible):
   Go:         https://docs.minio.io/docs/golang-client-quickstart-guide
   Java:       https://docs.minio.io/docs/java-client-quickstart-guide
   Python:     https://docs.minio.io/docs/python-client-quickstart-guide
   JavaScript: https://docs.minio.io/docs/javascript-client-quickstart-guide
   .NET:       https://docs.minio.io/docs/dotnet-client-quickstart-guide
^CExiting on signal interrupt
```

Reply:
@kannappanr you have to update to latest `cli` package

Reply:
Tested by locally vendorizing the latest version of cli package and the behavior does not change.